### DISCUSSION TITLE
Is there any reason we need the `nCycles` case setting?

### DISCUSSION THREAD

**Discussion by keckler** [terrapower/armi#579] [General]
`nCycles` seems redundant to me. A quick search through the code base makes me think that we could easily just infer the value from the various cycle history case settings.

But before I advocate getting rid of it entirely, does anybody see a reason why it needs to exist as its own setting? I think it should still be kept internally to avoid having to do something like `len(cycleLengths)` anytime we need the number of cycles, but it could just be inferred at the start of a run and stored in the case settings data structure where it currently is.

**Comment by john-science** [comment]
I am curious what others think of this. Off the top of my head, I don't see any good reason not to remove it. 

* The easiest solution would be to immediately calculate it at the start of a run. This allows us to change the least amount of code in ARMI.
* But if it really is just `len(cycleLenths)`, and we can say 100% of the time people will have `cycleLenghts`, I am actually okay with a large PR where we just remove it.

Do we even validate that `nCycles` and `cycleLengths` agree right now?  If not, this is another reason why reducing the redundancy of our inputs is a good idea.

**Reply by john-science** [reply]
Actually, after talking to people, it seems like `nCycles` is convenient for most people. It's nice to be able to see at a glance how many cycles you have, and if you have a `cycleLengths` in the dozens or hundreds of items long, counting them by hand is tedious. And a lot of people are currently defining their run by using configurations like `nCycles: 12` and `cycleLength: 2000.0`.  In this use case, having `nCycle` is consider quite convenient.

**Reply by keckler** [reply]
That is fine. I see why someone might want to enter input in that manner.

I guess what bothers me is that we force the input file to have redundant information by checking that `availabilityFactors`, `powerFractions`, and `cycleLengths` all have lengths that match `nCycles` through the use of `_factorsAreValid`:
https://github.com/terrapower/armi/blob/035711afd1514ec5914c59c3cf898dfb3213c5a4/armi/operators/settingsValidation.py#L465-L504

If we could optionally resolve `nCycles` at run time, for cases where it can be inferred, and change the `_factorsAreValid` check to instead just check that all three vectors are consistent lengths, that would be nicer, in my opinion.

**Reply by john-science** [reply]
Thoughts:

1. I'd be happy to change this method so that it allows for the check that all three have the same length in the case where `nCycles` isn't defined. Thus making `nCycles` potentially optional.
2. But we'd still need to do some work to ensure that `nCycles` is calculated on-the-fly for backwards compatibility reasons. This is, obviously, not hard to do. It would just be a matter of finding WHEN in the workflow to calculate it, to preserve our current functionality.

**Comment by ntouran** [comment]
Yeah I think things in the code should just check the internal `cycleLengths` data structure in all cases. It is still nice to have the user input option of using nCycles if you're inputting e.g. 300 cycles of the same cycle length for equilibrium analysis so I think we should retain the setting for that reason only. But all look-ups of the setting should be migrated to just check length of cycleLengths. Could make a property on the operator for convenience if we wanted.
